### DISCUSSION TITLE
enhanced leaflet support

### DISCUSSION THREAD

**Discussion by WolfgangFahl** [zauberzeug/nicegui#1446] [Show and tell]
see http://nicesrt.bitplan.com and https://github.com/WolfgangFahl/nicesrt for an enhanced leaflet support https://github.com/WolfgangFahl/nicesrt/blob/main/nicesrt/leaflet.py based on https://github.com/zauberzeug/nicegui/blob/main/examples/map/leaflet.py it add set_zoom_level and draw_path functions for a start.
<img width="882" alt="grafik" src="https://github.com/zauberzeug/nicegui/assets/1336221/d7320ca9-0267-4375-99d2-d1149dd59fe0">

**Comment by rodja** [comment]
Thanks for sharing @WolfgangFahl. Very interesting. We are also working on a map component because of some requirements from a few of our robot projects.

**Comment by WolfgangFahl** [comment]
@rodja  - I'd also need read access - how can i get e.g. the zoomlevel back via javascript? I'd love te get a layout similar to https://cycle.travel/map/journey/466345 with map and 2D/3D display 
![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/d5b7aa47-1322-41dc-8f30-3ea5d8e12723)

**Reply by falkoschindler** [reply]
Can't you hold props like `zoomlevel` in Python? So you don't need to read it back from the client, but simply return it in a Python property or method.

**Comment by WolfgangFahl** [comment]
The problem is that the code uses javascript functions that modify the zoomlevel. Currently only manual adjusting the slider to the same zoomlevel somewhat helps but that's IMHO only and ugly workaround. Generally methods should of course have return types. Even if the set of possible types would be limited e.g. only string or string/int/float/date that would already be helpful. Would this be a feature request with a new ticket?

**Reply by falkoschindler** [reply]
Yes, but why not keeping track of the current zoom level in Python like this:
```py
def set_zoom_level(self, zoom_level: int):
    self.zoom_level = zoom_level
    self.run_method("set_zoom_level", zoom_level)

def get_zoom_level(self) -> int:
    return self.zoom_level
```

Apart from that you're suggesting to add return values to `Element.run_method(...)`? Sure, that would probably be a useful feature and deserves its own ticket.

**Reply by WolfgangFahl** [reply]
@falkoschindler I do already keep track the zoom level set by nicetrack as suggested. So we only need the feature request.

**Reply by WolfgangFahl** [reply]
see #1471
